app/lib/report_helper.rb

```diff
@@ -1,23 +1,141 @@
+  UNUSED_IN_UI_FIELDS = ['genus_taxid', :sort_key].freeze
+
+  # use the metric's NT <=> NR dual as a tertiary sort key (so, for example,
+  # when you sort by NT R, entries without NT R will be ordered amongst
+  # themselves based on their NR R (as opposed to arbitrary ordder);
+  # and within the Z, for things with equal Z, use the R as tertiary
+  OTHER_COUNT_TYPE = {
+    'NT' => 'NR',
+    'NR' => 'NT'
+  }.freeze
+  OTHER_METRIC = {
+    'zscore' => 'r',
+    'r' => 'zscore',
+    'rpm' => 'zscore'
+  }.freeze
+
+  def threshold_param?(param_key)
+    parts = param_key.to_s.split "_"
+    (parts.length == 2 && parts[0] == 'threshold' && METRICS.include?(parts[1]))
+  end
+
+  def decode_thresholds(params)
+    thresholds = {}
+    METRICS.each do |metric|
+      param_key = "threshold_#{metric}".to_sym
+      thresholds[metric] = params[param_key]
+    end
+    thresholds
+  end
+
+  def decode_sort_by(sort_by)
+    parts = sort_by.split "_"
+    return nil unless parts.length == 3
+    direction = parts[0]
+    return nil unless SORT_DIRECTIONS.include? direction
+    count_type = parts[1].upcase
+    return nil unless COUNT_TYPES.include? count_type
+    metric = parts[2]
+    return nil unless METRICS.include? metric
+    {
+      direction:    direction,
+      count_type:   count_type,
+      metric:       metric
+    }
+  end
+
+  def number_or_nil(string)
+    Float(string || '')
+  rescue ArgumentError
+    nil
+  end
+
+  def valid_arg_value(name, value)
+    # return appropriately validated value (based on name), or nil
+    return nil unless value
+    if name == :view_level
+      value = value.downcase
+      value = nil unless VIEW_LEVELS.include? value
+    elsif name == :sort_by
+      value = nil unless decode_sort_by(value)
+    else
+      value = nil unless threshold_param?(name)
+      value = number_or_nil(value)
+    end
+    value
+  end
+
+  def clean_params(raw)
+    clean = {}
+    raw_hash = {}
+    raw.each do |name, value|
+      raw_hash[name] = value
+    end
+    raw = raw_hash
+    DEFAULT_PARAMS.each do |name, default_value|
+      raw_name = name.to_s
+      clean[name] = valid_arg_value(name, raw[raw_name]) || default_value
+      raw.delete(raw_name)
+    end
+    IGNORED_PARAMS.each do |name|
+      raw_name = name.to_s
+      raw.delete(raw_name)
+    end
+    logger.warn "Ignoring #{raw.length} report params: #{raw}." unless raw.empty?
+    clean
+  end
+
+  def external_report_info(report, params)
+    return {} if report.nil?
+    params = clean_params(params)
     data = {}
+    data[:report_page_params] = params
     data[:report_details] = report_details(report)
-    htc, td = taxonomy_details(report, params, view_level)
-    data[:highest_tax_counts] = htc
-    data[:taxonomy_details] = td
-    data[:view_level] = view_level
+    data[:taxonomy_details] = taxonomy_details(report, params)
     data[:all_categories] = all_categories
-    if params[:categories].present?
```

Comment:
If it's not too hard, can you add the following lines back? (even commented this fine)

Reply:
Yeah as part of reinstating category filter.   Thanks.